tests/ibeam/src/test_gateway_client.py

```diff
@@ -0,0 +1,220 @@
+    """
+    GatewayClient.secret_value tests both via init and when
+    directly called
+    """
+
+    # rnd generates a random value w/o any control
+    # characters
+    def rnd(n=16):
+        c = string.ascii_letters + string.digits + string.punctuation
+        return ''.join(random.choice(c) for i in range(n))
+
+    # basic tests to run include specifying the secrets
+    # directly, reading them from the environment, and
+    # reading them from the filesystem
+    tests = [
+        types.SimpleNamespace(
+            name='passed values',
+            source=None,
+            initial_env={
+            },
+        ),
+        types.SimpleNamespace(
+            name='env values w/o IBEAM_SECRETS_SOURCE',
+            source=SECRETS_SOURCE_ENV,
+            initial_env={
+            },
+        ),
+        types.SimpleNamespace(
+            name='env values w/ IBEAM_SECRETS_SOURCE',
+            source=SECRETS_SOURCE_ENV,
+            initial_env={
+                'IBEAM_SECRETS_SOURCE': SECRETS_SOURCE_ENV,
+            },
+        ),
+        types.SimpleNamespace(
+            name='fs values w/ IBEAM_SECRETS_SOURCE',
+            source=SECRETS_SOURCE_FS,
+            initial_env={
+                'IBEAM_SECRETS_SOURCE': SECRETS_SOURCE_FS,
+            },
+        ),
+    ]
+
+    # names of the secret values
+    ibeam_fields = [
+        'IBEAM_ACCOUNT',
+        'IBEAM_PASSWORD',
+        'IBEAM_KEY',
+    ]
+
+    for t in tests:
+        # GatewayClient init variables
+        account = None
+        password = None
+        key = None
+
+        # generate random secret values
+        for k in ibeam_fields:
+            setattr(t, k, rnd())
+
+        # setup test state based on the t.source value
+        if t.source is None:
+            # passing init values directly
+            account = t.IBEAM_ACCOUNT
+            password = t.IBEAM_PASSWORD
+            key = t.IBEAM_KEY
+
+            # noop test_setup
+            def test_setup(t, lappend='', rappend=''):
+                return mock.patch.dict(os.environ, {})
+
+        elif t.source == SECRETS_SOURCE_ENV:
+            # test_setup sets the source and copies the
+            # secrets into the environment
+            def test_setup(t, lappend='', rappend=''):
+                environ = t.initial_env
+                for k in ibeam_fields:
+                    environ[k] = lappend + getattr(t, k) + rappend
+                return mock.patch.dict(os.environ, environ)
+        elif t.source == SECRETS_SOURCE_FS:
+            # test_setup sets the source and copies the
+            # secrets onto the filesystem, setting the
+            # filepath into the environment
+            def test_setup(t, lappend='', rappend=''):
+                environ = t.initial_env
+                for k in ibeam_fields:
+                    environ[k] = os.path.join(tmpdir, k)
+                    with open(environ[k], 'wt', encoding='UTF-8') as fh:
```

Comment:
Is this actually reading a file? Do we have these files present?

Reply:
It's writing out the random secrets we generated earlier out to files on the temporary directory pytest manages.

Reply:
Great, thanks for explaining this 👍